### ISSUE TITLE
@Blocking being ignored when accessing entity stream in a request filter

### ISSUE THREAD

**Issue by nikosk686** [quarkusio/quarkus#32359]
### Describe the bug

Trying to access the `ContainerRequestContext` entity stream from within a `ServerRequestFilter` fails even if the whole flow is annotated with `@Blocking`.

### Expected behavior

The the blocking operation of parsing the request body as an entity stream should not fail when `@Blocking` is used.

### Actual behavior

The request fails with the following errors
```
2023-04-03 12:56:21,844 ERROR [org.jbo.res.rea.ser.cor.RuntimeExceptionMapper] (vert.x-eventloop-thread-0) A blocking operation occurred on the IO thread. This likely means you need to use the @io.smallrye.common.annotation.Blocking annotation on the Resource method, class or javax.ws.rs.core.Application class.
2023-04-03 12:56:21,848 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-0) HTTP Request to /hello failed, error id: 23f3890e-eda5-4455-9340-34ef39fff0cd-1: org.jboss.resteasy.reactive.common.core.BlockingNotAllowedException: Attempting a blocking read on io thread
```

### How to Reproduce?

Execute this test for the following resource
```
    @Test
    public void testHelloEndpoint() {
        given()
          .when()
                .header("Content-Type", "application/json")
                .body("{\"name\": \"me\"}")
                .post("/hello")
          .then()
             .statusCode(200)
             .body(is("Hello from me"));
    }
```
```
@Blocking
@Path("/hello")
public class ExampleResource {

    @ServerRequestFilter(preMatching = true)
    public void filter(ContainerRequestContext requestContext) {
        String body = new BufferedReader(new InputStreamReader(requestContext.getEntityStream()))
                .lines().collect(Collectors.joining("\n"));
        requestContext.setEntityStream(new ByteArrayInputStream(body.getBytes()));
    }


    @POST
    @Produces(MediaType.TEXT_PLAIN)
    @Consumes(MediaType.APPLICATION_JSON)
    public Uni<String> hello(Map<String, String> body) {
        return Uni.createFrom().item("Hello from %s".formatted(body.get("name")));
    }
}
```

### Output of `uname -a` or `ver`

Linux tymit-XPS-15-9500 5.19.0-38-generic #39~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Mar 17 21:16:15 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

OpenJDK 64-Bit Server VM Corretto-17.0.5.8.1 (build 17.0.5+8-LTS, mixed mode, sharing)

### Quarkus version or git rev

2.16.5.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.5.1

### Additional information

Related issues:

1. https://github.com/quarkusio/quarkus/issues/30312
2. https://github.com/quarkusio/quarkus/issues/17280

**Comment by geoand** [comment]
Hm, this is actually kind of a weird case because it combines the use of a pre-matching filter with a per-resource declaration, which doesn't make a tremendous amount of sense (and I would argue should probably be made an error).

@nikosk686 what's the reason for making the filter a pre-matching one?

**Comment by nikosk686** [comment]
Thanks for the feedback @geoand.
I can confirm that in a **non** `pre-matching` filter the `@Blocking` annotation works as expected in my above example.
So while I have a reason to use `pre-matching` in my case, which I ll explain shortly in the sake of clarity, even in my case this is more like a hacky workaround attempt rather than a valid case.
Said that feel free to close the present ticket if you consider that there are not any affected valid use cases justifying the effort of taking care of this.

So why I was trying with a `pre-matching` filter?
I am experimenting with the not yet released `@PermissionsAllowed` [annotation](https://github.com/quarkusio/quarkus/pull/31345) and in my case I need to involve a [SecurityIdentityAugmentor](https://quarkus.io/guides/security-customization#security-identity-customization) which will assign some `Permission`s to the `SecurityIdentity` object in order to be checked against the value of the `@PermissionsAllowed`.
The assigned permissions will be defined driven by a specific value in the request body and that's why I am trying to extract it in the filter (in order to create a custom `SecurityContext` including the extracted informantion and pass it to the `ContainerRequestContext` so it can be used by my augmentor).
Until now I haven't find a way to execute the filter first, before the augmentor, and the `@Priority` annotation didn't help much here, unless I am doing something obviously wrong - still checking.
So in any case the augmentor is being executed first. But without the correct permission assigned the next step of the flow is to get a 403 error response. With a pre-matching filter instead the flow is getting into the filter before the 403 response, something that allowes me to do some more debugging in the attempt to achieve what I am trying to do. This is how I noticed the currently reported issue.
But as I said above feel free to close it as even in my weird case, it wouldn't solve the problem I am facing with the execution order.

**Comment by michalvavrik** [comment]
> Until now I haven't find a way to execute the filter first, before the augmentor, and the @Priority annotation didn't help much here, unless I am doing something obviously wrong - still checking.

Yes, you will never find it unless proactive authentication is disabled and you don't use HTTP security policy etc. (you need filters to execute before authentication). It is also documented that permissions can't work with adjusted `SecurityContext` because there are no permissions, please see https://quarkus.io/version/main/guides/security-authorize-web-endpoints-reference.

> The assigned permissions will be defined driven by a specific value in the request body

while this actually please me to hear (as it validates use case for passing resource method arguments to the permission), I think it also says 2 things:

- you are going to be blocked by https://github.com/quarkusio/quarkus/issues/32030, I'll try to look at it in 2 weeks, but TBH I don't have too much time to work on Quarkus now, let's subscribe to the issue and watch notifications
- I'd suggest you can turn logic around - you can determine if identity should be allowed to pass in the method permission that gets passed in resource method args. Just assign to authenticated user roles according to the credentials some permission and decide whether it should pass in the method permission.

After all that is done, I'll try to improve docs to make it clearer on first read, but it has to wait now.
